python/tank/authentication/user.py

```diff
@@ -118,6 +128,109 @@ def impl(self):
+
+    This specialized class allows the user to query the claims expiration and
+    see if `interactive_authentication.renew_session(user)` needs to be called.
+
+    It is also possible to start/stop/query the state of the automatic claims renewal:
+
+        user = ...
+        if isinstance(user, ShotgunSamlUser):
+            user.start_claims_renewal()
+        ...
+        if user.is_claims_renewal_active():
+            user.stop_claims_renewal()
+    """
+
+    def __init__(self, impl):
+        """
+        :param impl: Internal user implementation class this class proxies.
+        """
+        super(ShotgunSamlUser, self).__init__(impl)
+        self._timer = None
+
+    def get_claims_expiration(self):
+        """
+        Obtain claims expiration for the user.
+
+        :returns: The expiration in seconds since January 1st 1970 UTC.
+        """
+        # The import must be located here. If it is at global scope, there
+        # is an import error.
+        from .shotgun_shared import saml2_sso
+
+        return saml2_sso.get_saml_claims_expiration(self._impl.get_cookies())
+
+    def _automatic_claims_renewal(self, preemtive_renewal_threshold=0.9):
+        """
+        Handles automatic renewal of the SAML2 claims for the user.
+
+        :params user: an already logged in user.
+        :params preemtive_renewal_threshold: How far into the claims duration we will attempt renewal.
+                                             Defaults to 90%, usually 3 minutes 45 seconds (90% of 5 mins).
+        """
+        logger.debug("Automatic claims renewal")
+        try:
+            previous_expiration = self.get_claims_expiration()
+            # A call to renew_session when SSO is used will not prompt the user for
+            # their credentials if it is not necessary.
+            interactive_authentication.renew_session(self._impl)
+            new_expiration = self.get_claims_expiration()
+
+            if new_expiration > previous_expiration:
+
+                logger.debug("Automatic claims renewal succeeded.")
+                delta = (new_expiration - time.time()) * preemtive_renewal_threshold
+                # If we are debugging, we will use a shorter expiration time.
+                # SHOTGUN_SSO_RENEWAL_INTERNAL should be a value in seconds.
+                if "SHOTGUN_SSO_RENEWAL_INTERNAL" in os.environ:
+                    delta = int(os.environ['SHOTGUN_SSO_RENEWAL_INTERNAL'])
+                logger.debug("Next claims renewal attempt: %f" % delta)
+                self._timer = threading.Timer(delta, self._automatic_claims_renewal, [preemtive_renewal_threshold])
+                self._timer.start()
+            else:
+                logger.warning("Automatic claims renewal failed. No longer attempting to renew claims.")
+        except AuthenticationCancelled:
+            logger.debug("Authentication cancelled, most likely from quitting the application.")
```

Comment:
rather than guessing, let's just report what we are seeing e.g. something like `Automatic SSO claim renewal was cancelled while processing.` The rest of the debug log will provide ample context around what is happening (quitting vs other stuff)